rfcs/0148-pipe-operator.md

```diff
@@ -0,0 +1,223 @@
+Having a dedicated operator also increases visibility and discoverability of the feature.
+
+# Detailed design
+[design]: #detailed-design
+
+## `|>` operator
+
+A new operator `|>` is introduced into the Nix language.
+Semantically, it is defined as the reverse of function application: `f a` = `a |> f`.
+It is left-associative and has a binding strength one weaker than function application:
+`a |> f |> g b |> h` = `h ((g b) (f a))`.
+
+## `builtins.pipe`
+
+`lib.pipe`'s functionality is implemented as a built-in function.
+The main motivation for this is that it allows to give better error messages
+like line numbers when some part of the pipeline fails.
+Additionally, it allows easy usage outside of Nixpkgs and increases discoverability.
+
+While Nixpkgs is bounds to minimum Nix versions and thus `|>` won't be available until
+several years after its initial implementation,
+it can directly benefit from `builtins.pipe` and its better error diagnostic by overriding `lib.pipe`.
+Elevating a Nixpkgs library function to a builtin has been done several times before,
+for example `bitAnd`, `splitVersion` and `concatStringsSep`.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+## Tooling support
+
+Like any language extension, this will require the available Nix tooling to be updated.
+Updating parsers should be pretty easy, as the syntax changes to the language are fairly minimal.
+Tooling that evaluates Nix code in some way or does static code analysis should be easy to support too,
+since one may treat the operator as syntactic sugar for function application.
+No fundamentally new semantics are introduced to the language.
+
+# Prior art
+
+Nickel has `|>` too, with the same name and semantics.
+
+F# has `|>`, called "pipe-forward" operator, with the same semantics.
+Additionally, it also has "pipe-backward" `<|` and `>>`/`<<` for forwards and backwards function composition.
+`<|` is equivalent to function application, however its lower binding order allows removing parentheses:
+`g (f a)` = `g <| f a`. All these operators have the same precedence and are left-associative.
+
+Elm has the same operators as F#.
+
+Haskell has the (backwards) function composition operator `.` in its prelude: `(g . f) a` = `g (f a)`.
+It also has "reverse application" `&`, which is roughly equivalent to `|>`,
+and `$`, which is function application again but right-associative and very weakly binding.
+`.` binds stronger than both.
+
+`|>` is definable as an infix function in several other programming languages,
+and in even more languages as macro or higher-order function (including Nix, that's `lib.pipe`).
+
+# Alternatives
+[alternatives]: #alternatives
+
+For each change this RFC proposes, there is always the trivial alternative of not doing it. See #drawbacks.
+
+## More operators
+
+We could use the occasion and introduce more operators like those mentioned above.
```

Comment:
I think introducing the pipe operator is hard enough as it is. While using the occasion to also merge other operators makes sense, I don't see the future where it happens, given the organizational difficulties and us not having any mechanism to do that without treewide breakages: we don't have language versions.

Reply:
Shepherds currently believe this conversation is resolved, outcome: inclined to leave `pipe` in `lib`.